### PR TITLE
Add GitLab code quality reporting.

### PR DESCRIPTION
**Pull request by cmlara**
This PR:

- [x] Adds new feature ...
- [x] Covered by tests
- [x] Doc PR: https://github.com/infection/site/pull/254

Fixes https://github.com/infection/infection/issues/1877

Adds a new logger format that outputs the GitLab(Code Climate) code quality format.

### CODE REVIEW

File: resources/schema.json [hunk 1]
```diff
@@ -96,6 +96,10 @@
                     "type": "boolean",
                     "description": "GitHub Annotations for escaped Mutants in the added/modifies files"
                 },
+                "gitlab": {
+                    "type": "string",
+                    "definition": "GitLab(Code Climate) code quality json file."
```
Comment (maks-rafalko) [inline resources/schema.json:101]:
```suggestion
                    "definition": "GitLab (Code Climate) code quality json file."
```

?

File: src/Configuration/ConfigurationFactory.php [hunk 2]
```diff
@@ -327,6 +327,7 @@ private function retrieveLogs(Logs $logs, string $configDir, ?bool $useGitHubLog
             self::pathToAbsolute($logs->getHtmlLogFilePath(), $configDir),
             self::pathToAbsolute($logs->getSummaryLogFilePath(), $configDir),
             self::pathToAbsolute($logs->getJsonLogFilePath(), $configDir),
+            self::pathToAbsolute($logs->getGitlabLogFilePath(), $configDir),
```
Comment (maks-rafalko) [inline src/Configuration/ConfigurationFactory.php:330]: I think we also need `--logger-gitlab` or something like this, to 

- make it consistent with `--logger-github`
- to have option and config file parity, where this feature can be enabled either from option (`--logger-gitlab`) or from config file by setting `"github": "/path/to/file.json"`

Related to https://github.com/infection/infection/issues/141

**Review by maks-rafalko — COMMENTED** [review]
Thank you for this great addition!

Left a couple of comments and we are good to go.

I will ask you also update the docs and add information about new option and config setting, see examples:

- https://infection.github.io/guide/command-line-options.html#logger-github
- https://infection.github.io/guide/usage.html

Repository to update: https://github.com/infection/site/blob/master/src/guide/command-line-options.md

**Comment by cmlara** [comment]
Created PR for documentation and added the CLI option as requested.

**Review by maks-rafalko — APPROVED** [review]
Thank you for your work, much appreciated.

**Comment by maks-rafalko** [comment]
Released, https://github.com/infection/infection/releases/tag/0.27.3

If you have some public repository where I can see this in action, please let me know.

Thanks.

**Comment by cmlara** [comment]
Thank you! 

This was the first step in working towards including infection in a project workflow, I will circle back with an example once I have it on a public repo.

**Comment by cmlara** [comment]
Here are a couple reference links on a public GitLab server showing it in action configured to perform full scans on branch CI runs and to only scan modified files for MR's.

MR Overview showing the mutation reported as a Code Quality fault: https://git.drupalcode.org/project/s3fs/-/merge_requests/36
MR Changes page showing the escaped mutant as an expandable in-line: https://git.drupalcode.org/project/s3fs/-/merge_requests/36/diffs#0b97329dffce8550a1ce638084c975392cad49c9_49_50

**Comment by maks-rafalko** [comment]
@cmlara I'm new to GitLab, so probably not understand somthing completely, but do you know why it doesn't work on my very simple minimal example?

https://gitlab.com/maks-rafalko/infection-gitlab-integration-example/-/merge_requests/2

1. I don't see any "annotations" as on yours PR. Is it because I use free account?
2. I don't see any files (artifacts) uploaded to PR, only can found them on job details page. Is it expected?
3. Even [those uploaded artifacts are empty](https://maks-rafalko.gitlab.io/-/infection-gitlab-integration-example/-/jobs/5294006126/artifacts/infection-gitlab.json), why Infection generated 2 escaped mutations

Pipeline definition: https://gitlab.com/maks-rafalko/infection-gitlab-integration-example/-/blob/main/.gitlab-ci.yml#L25-30

**Comment by cmlara** [comment]
> I don't see any "annotations" as on yours PR. Is it because I use free account?

Unfortunately the "See findings in merge request diff view" does require an Ultimate account which is why it works 

>I don't see any files (artifacts) uploaded to PR, only can found them on job details page. Is it expected?

On https://gitlab.com/maks-rafalko/infection-gitlab-integration-example/-/merge_requests/2 from the download icon they are in the "infection:archive" and "code_quality:archive" download options (GitLab batches all the artifacts into a single download) the filename can be changed by adding an "expose_as" key https://docs.gitlab.com/ee/ci/yaml/#artifactsexpose_as

>Even those uploaded artifacts are empty, why Infection generated 2 escaped mutations

On quick glance it does not look like it is a configuration fault in your gitlab-ci.yml. I'm going to have to load up a lab to debug to see if there is a fault in how I implemented the logger plugin.

**Comment by cmlara** [comment]
Interesting: 

When I add on the HTML logger it appears to works:

./vendor/bin/infection --logger-gitlab=infection-gitlab.json --logger-html=/dev/null 
(I used https://github.com/firecow/gitlab-ci-local to run your repo locally) 

I'm not very familiar with infection's internals, any thoughts on what might cause this? I don't see this on another project, maybe its a dependency related fault?

**Comment by maks-rafalko** [comment]
> I'm not very familiar with infection's internals, any thoughts on what might cause this?

there is a bug, fix is here: https://github.com/infection/infection/pull/1882

> I don't see this on another project, maybe its a dependency related fault?

I'm pretty sure that when Infection is used **only** with `--logger-gitlab` - it can't work correctly. Can you share a project and/or executable command line?

**Comment by maks-rafalko** [comment]
Now it works on my sample project!

https://gitlab.com/maks-rafalko/infection-gitlab-integration-example/-/merge_requests/2

image

**Comment by cmlara** [comment]
Ah I missed a location when adding the feature, thank you for finding and creating the fix.

To followup on how I failed to reproduce earlier: Looking back at my command logs it appears I was testing with the -s flag that would also be a triggering option. This does appear to be in line with the fix you provided.

**Pull request merged**